src/wrapper/objects/global_ref.rs

```diff
@@ -1,143 +1,104 @@
-            obj: self.obj,
-            vm: vm,
-        };
-
-        mem::forget(self); // prevent dropping the reference.
-
-        Ok(res)
-    }
-}
-
-impl Drop for GlobalRef {
-    fn drop(&mut self) {
-        let res = self.drop_ref();
-        match res {
-            Ok(()) => {}
-            Err(e) => debug!("error dropping global ref: {:#?}", e),
-        }
+        self.inner.as_obj()
     }
 }
 
-/// A detached global JVM reference that can be sent across threads. To do
-/// anything useful with it, it must be `attach`ed first.
-pub struct DetachedGlobalRef {
-    obj: JObject<'static>,
-    vm: JavaVM,
-}
-
-unsafe impl Send for DetachedGlobalRef {}
 
-impl DetachedGlobalRef {
-    /// Creates a new detached global reference. This assumes that
-    /// `CreateGlobalRef`
-    /// has alrady been called.
-    pub unsafe fn new(vm: JavaVM, obj: sys::jobject) -> Self {
-        DetachedGlobalRef {
+impl GlobalRefGuard {
+    /// Creates a new global reference. This assumes that `NewGlobalRef`
+    /// has already been called.
+    unsafe fn new(vm: JavaVM, obj: sys::jobject) -> Self {
+        GlobalRefGuard {
             obj: JObject::from(obj),
             vm,
         }
     }
 
-    /// Attach this ref to a `JNIEnv` to produce `GlobalRef`.
-    pub fn attach(self, env: &JNIEnv) -> GlobalRef {
-        let res = self.attach_impl(env);
-        mem::forget(self);
-        res
-    }
-
-    /// Unwrap to the internal JNI type.
-    pub fn into_inner(self) -> sys::jobject {
-        self.obj.into_inner()
+    /// Get the object from the global ref
+    ///
+    /// This borrows the ref and prevents it from being dropped as long as the
+    /// JObject sticks around.
+    pub fn as_obj<'a>(&'a self) -> JObject<'a> {
+        self.obj
     }
+}
 
-    fn drop_impl(&self) -> Result<()> {
-        match self.vm.get_env() {
-            Ok(env) => {
-                let _ = self.attach_impl(&env);
-                Ok(())
-            }
-            Err(Error(ErrorKind::ThreadDetached, _)) => {
-                let env = self.vm.attach_current_thread()?;
-                let _ = self.attach_impl(&env);
-                Ok(())
+impl Drop for GlobalRefGuard {
+    fn drop(&mut self) {
+        fn drop_impl(env: &JNIEnv, global_ref: JObject) -> Result<()> {
+            let internal = env.get_native_interface();
+            unsafe {
+                jni_unchecked!(internal, DeleteGlobalRef, global_ref.into_inner());
+                check_exception!(internal);
             }
-            Err(e) => Err(e),
+            Ok(())
         }
-    }
 
-    fn attach_impl(&self, env: &JNIEnv) -> GlobalRef {
-        GlobalRef {
-            obj: self.obj,
-            env: env.get_native_interface(),
-        }
-    }
-}
+        let res = match self.vm.get_env() {
+            Ok(env) => drop_impl(&env, self.as_obj()),
+            Err(_) => {
+                warn!("`GlobalRef` attaches JNI-thread while dropping its instance.");
```

Comment:
Since this message is for (potentially) ignorant users, I'd be more verbose and suggest what they shall do:
`Dropping a GlobalRef in a detached thread. Fix your code if this message appears frequently (see the GlobalRef docs)`

Currently it's unclear from the message whether it shall concern a user at all.